File: src/main/java/eu/mshade/enderframe/tick/Tickable.java
package eu.mshade.enderframe.tick;

public abstract class Tickable {

    private long tick;

    private TickBus tickBus;

    private TickState tickState = TickState.WAITING;


    public abstract void tick();

    public boolean isPeriod(int period){
        return tick % period == 0;
    }

    public void addTick(){
        this.tick++;
    }

    public long getTick() {
        return tick;
    }

    public void resetTick(){
        this.tick = 0;
    }
    public TickState getTickState() {
        return tickState;
    }

    public void setTickState(TickState tickState) {
        this.tickState = tickState;
    }

    public void joinTickBus(TickBus tickBus){
        this.tickBus = tickBus;
        tickBus.addTickable(this);
    }

    public void leaveTickBus(){
        if (this.tickBus != null) {
            tickBus.removeTickable(this);
            tickBus = null;
        }
    }
}

File: src/main/java/eu/mshade/enderframe/tick/Tickable.kt
package eu.mshade.enderframe.tick

abstract class Tickable {

    var tick: Long = 0
        private set
    val tickId = TickBus.TICKABLE_ID.freeId

    private var tickBus: TickBus? = null
    var tickState = TickState.WAITING

    abstract fun tick()

    fun isPeriod(period: Int): Boolean {
        return tick % period == 0L
    }

    fun addTick() {
        tick++
    }

    fun resetTick() {
        tick = 0
    }

    fun joinTickBus(tickBus: TickBus) {
        if (this.tickBus != null) return
        this.tickBus = tickBus
        tickBus.addTickable(this)
    }

    fun leaveTickBus() {
        if (tickBus != null) {
            tickBus!!.removeTickable(this)
            tickBus = null
        }
    }
}
